### ISSUE TITLE
Get read names from mpileup

### ISSUE THREAD

**Issue by bdeonovic** [samtools/samtools#473]
Is there a way to get the read names of the reads associated with an mpileup? For example mpileup can give me the following output:

```bash
seq1 272 T 24  ,.$.....,,.,.,...,,,.,..^+. <<<+;<<<<<<<<<<<=<;<;7<&
seq1 273 T 23  ,.....,,.,.,...,,,.,..A <<<;<<<<<<<<<3<=<<<;<<+
```

is there a way to have it output the readnames associated with those nucleotide/quality calls?

**Comment by jkbonfield** [comment]
The answer unfortunately is no.  Adding additional fields would break too many downstream applications, so we'd need a very compelling reason to change this.

However the htslib API itself serves up mpileup line by line as an array of bam structures, from which anything and everything can be fetched if you wish.  While this probably isn't the answer you're hoping for, take a look at  bam_plcmd.c#L553 (//begin pileup) to see how the main loop works.

Basically you have to initialise a bam_mplp_t iterator with bam_mplp_init and then bam_mplp_auto returns successive columns one at a time.  Sorry we can't help much more than this.

**Comment by sedlawrence** [comment]
Do the calls on each line correspond to each other? i.e. does the first comma on each line correspond to the same read? If not, is there any way to retrieve this info?

**Comment by jkbonfield** [comment]
The order of elements in the sequence column correspond to the same order in the qual column, and read name columns, etc.  

Reads that terminate have a `$` marker in the sequence and reads that appear have `^`<mapping qual> characters.  These are typically the start and end of the seq column respectively, although there's no explicit guarantee of such in the documentation so I wouldn't rely on consistency.  Also note that with base quality filtering reads and appear and disappear multiple times, including potentially the never showing a `^` or `$` symbol as the first and last base got filtered.  Basically, read the format description in the [man page](http://www.htslib.org/doc/samtools-mpileup.html) for the full details.

It's definitely not the easiest format in the world to parse, but there are options which can remove some of the harder bits if you don't care about those.  Also, if you find yourself spending a lot of time doing text parsing, you're probably better off coding it up in C or perhaps python via PySam.  See https://github.com/samtools/htslib/blob/develop/test/pileup.c for a simple example of pileup from C.

**Comment by camelest** [comment]
@jkbonfield Hi, thank you so much for the following up on this. I'm also facing the same problem.

I went through the manual and regarding the point:
> there are options which can remove some of the harder bits if you don't care about those. 

is there any other options I could try except `--no-output-ends` ? With this, the column4 in the output (Number of reads covering this position) still does not match the strings in the column5 (Read bases). 

```shell
samtools mpileup -r <position> <input_primary_alignment_only.bam> --output-extra QNAME -f <reference>  --no-output-ends > reads_noends.pileup

awk '{print $3}' reads_primary_noends.pileup
4444

awk '{print $5}' reads_primary_noends.pileup | wc
1 1 4451

awk '{print $5}' reads_primary_noends.pileup |  awk '{split($0, chars, ""); for (i in chars) count[chars[i]]++} END {for (char in count) print char ": " count[char]}'`
A: 1
*: 4
+: 1
,: 2747
-: 1
a: 3
G: 13
.: 38
<: 21
1: 2
>: 7
g: 1612
``` 

I also tried `--no-BAQ` but still saw some discrepancy.

Thank you so much for the help.

**Comment by jkbonfield** [comment]
That looks like you're getting mapping qualities in there, and probably indels too (the `+` and `-` indicates that).

See the usage for an idea of extra options:

```
Output options:
  -o, --output FILE        write output to FILE [standard output]
  -O, --output-BP          output base positions on reads, current orientation
      --output-BP-5        output base positions on reads, 5' to 3' orientation
  -M, --output-mods        output base modifications
  -s, --output-MQ          output mapping quality
      --output-QNAME       output read names
      --output-extra STR   output extra read fields and read tag values
      --output-sep CHAR    set the separator character for tag lists [,]
      --output-empty CHAR  set the no value character for tag lists [*]
      --no-output-ins      skip insertion sequence after +NUM
                           Use twice for complete insertion removal
      --no-output-ins-mods don't display base modifications within insertions
      --no-output-del      skip deletion sequence after -NUM
                           Use twice for complete deletion removal
      --no-output-ends     remove ^MQUAL and $ markup in sequence column
      --reverse-del        use '#' character for deletions on the reverse strand
  -a                       output all positions (including zero depth)
  -a -a (or -aa)           output absolutely all positions, including unused ref. sequences
```

So you probably want all the `--no-output` options if you just want bases and nothing else.
